src/tools/wasm-fuzz-lattices.cpp

```diff
@@ -0,0 +1,479 @@
+  AnalysisChecker<FiniteIntPowersetLattice, LivenessTransferFunction> checker;
+  LivenessChecker(Function* func, uint64_t latticeElementSeed, Name funcName)
+    : lattice(func->getNumLocals()), checker(lattice,
+                                             transferFunction,
+                                             "FiniteIntPowersetLattice",
+                                             "LivenessTransferFunction",
+                                             latticeElementSeed,
+                                             funcName) {}
+
+  FiniteIntPowersetLattice::Element getRandomElement(Random& rand) {
+    FiniteIntPowersetLattice::Element result = lattice.getBottom();
+
+    // Uses rand to randomly select which members are to be included (i. e. flip
+    // bits in the bitvector).
+    for (size_t i = 0; i < lattice.getSetSize(); ++i) {
+      result.set(i, rand.oneIn(2));
+    }
+    return result;
+  }
+};
+
+// Struct to set up and check reaching definitions analysis lattice and transfer
+// function.
+struct ReachingDefinitionsChecker {
+  LocalGraph::GetSetses getSetses;
+  LocalGraph::Locations locations;
+  ReachingDefinitionsTransferFunction transferFunction;
+  AnalysisChecker<FinitePowersetLattice<LocalSet*>,
+                  ReachingDefinitionsTransferFunction>
+    checker;
+  ReachingDefinitionsChecker(Function* func,
+                             uint64_t latticeElementSeed,
+                             Name funcName)
+    : transferFunction(func, getSetses, locations),
+      checker(transferFunction.lattice,
+              transferFunction,
+              "FinitePowersetLattice<LocalSet*>",
+              "ReachingDefinitionsTransferFunction",
+              latticeElementSeed,
+              funcName) {}
+
+  FinitePowersetLattice<LocalSet*>::Element getRandomElement(Random& rand) {
+    FinitePowersetLattice<LocalSet*>::Element result =
+      transferFunction.lattice.getBottom();
+
+    // Uses rand to randomly select which members are to be included (i. e. flip
+    // bits in the bitvector).
+    for (size_t i = 0; i < transferFunction.lattice.getSetSize(); ++i) {
+      result.set(i, rand.oneIn(2));
+    }
+    return result;
+  }
+};
+
+struct Fuzzer {
+  bool verbose;
+
+  Fuzzer(bool verbose) : verbose(verbose) {}
+
+  // Helper function to run per-function tests. latticeElementSeed is used to
+  // generate three lattice elements randomly. It is also used to select which
+  // analysis is to be tested for the function.
+  void runOnFunction(Function* func, uint64_t latticeElementSeed) {
+    RandEngine getFuncRand(latticeElementSeed);
+
+    // less bytes are needed to generate three random lattices.
+    std::vector<char> funcBytes(128);
+    for (size_t i = 0; i < funcBytes.size(); i += sizeof(uint64_t)) {
+      *(uint64_t*)(funcBytes.data() + i) = getFuncRand();
+    }
+
+    Random rand(std::move(funcBytes));
+
+    CFG cfg = CFG::fromFunction(func);
+
+    switch (rand.upTo(2)) {
+      case 0: {
```

Comment:
Maybe each of these cases can be packaged up into a separate function, just to keep the `runOnFunction` routine shorter and easier to read.

Reply:
I could probably consolidate this under `AnalysisChecker` entirely if I passed a pointer to one the specific checker to it. However, I don't know if that would be too messy, since the `AnalysisChecker` is already the owner of the specific checker.

Reply:
This LGTM how it is 👍